[quant][graph] Add quant fusion for dynamic quantization

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35586 [quant][graph] Add quant fusion for dynamic quantization**
* #35448 [quant][graph] Add pass to insert quant dequant for dynamic quantization

Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755680](https://our.internmc.facebook.com/intern/diff/D20755680)

supriyar committed:
[quant][graph] Add quant fusion for dynamic quantization
Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/test_quantize_script.py:
```diff
@@ -168,5 +170,21 @@ def forward(self, x):
                    .check("return") \
                    .run(str(get_forward_graph(m.fc1._c)))
 
+    def test_finalize_for_linear_dynamic(self):
```
should be: dynamic_linear

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -2485,7 +2485,15 @@ void SwapDeQuant(std::shared_ptr<Graph>& graph) {
+      rewriter.runOnGraph(graph);
+    }
+    return;
+  }
```
could you use the same code?
Also I think you'll need to rebase, return type for quant_fusion_pattern_and_replacements is changed

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -115,3 +123,16 @@ def quantize_script(model, qconfig_dict, run_fn, run_args, inplace=False, debug=
     model = convert_script(model, True, debug)
     return model
+
+def quantize_dynamic_script(model, qconfig_dict, run_fn, run_args, debug=False):
```
could you reuse the `quantize_script` code?

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -95,6 +95,14 @@ def convert_script(model, inplace=False, debug=False):
         model = wrap_cpp_module(torch._C._jit_pass_quant_finalize(model._c))
     return model
 
+def convert_dynamic_script(model, debug=False):
```
same here, could you reuse the code?

jerryzh168 reviewed (changes requested):
I think we should reuse most of the code

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -95,26 +99,44 @@ def prepare_dynamic_script(model, qconfig_dict):
+    if is_dynamic:
+        model = prepare_dynamic_script(model, qconfig_dict)
+    else:
+       model = prepare_script(model, qconfig_dict, True)
```
identation

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -95,26 +99,44 @@ def prepare_dynamic_script(model, qconfig_dict):
     return model
 
-def convert_script(model, inplace=False, debug=False):
+def _convert_script(model, is_dynamic, debug=False):
```
could you refactor prepare as well?

jerryzh168 reviewed (changes requested):
please fix indentations, there are a few places were indentations are wrong

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755680](https://our.internmc.facebook.com/intern/diff/D20755680)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755680](https://our.internmc.facebook.com/intern/diff/D20755680)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755680](https://our.internmc.facebook.com/intern/diff/D20755680)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/_quantize_script.py:
```diff
@@ -74,47 +78,61 @@ def script_qconfig(qconfig):
+    if is_dynamic:
+        model = convert_dynamic_script(model, debug)
+    else:
+        model = convert_script(model, True, debug)
```
could you merge the two ifs?

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -1081,6 +1081,43 @@ def test_skip_quant(self):
+
+            # Check to make sure choose_qparams->quant->dequant->linear is numerically
+            # equivalent to the final quantized model.
+            model_fake_quantized = quantize_dynamic_script(
```
you mean model_dynamic_quantized?

jerryzh168 replied:
Oh, this is debug, could you use another name? fake_quantize has a special meaning in quantization

supriyar replied:
No, this tests the fake quantized model with the `quantize_per_tensor` op followed by float Linear op to be numerically equivalent to the dynamically quantized op in eager mode.

jerryzh168 reviewed (approved):
LG, some comments are inline

supriyar committed:
Update on "[quant][graph] Add quant fusion for dynamic quantization"


Summary:
This pass fuses the choose_qparams-quant-dequant sequence
Fusion for weight tensor is the same as static quant.

Test Plan:
python test/test_quantize_script.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20755680](https://our.internmc.facebook.com/intern/diff/D20755680)

[ghstack-poisoned]
